### PR TITLE
fix(ports): ban conflicting port ranges for colocated units

### PR DESCRIPTION
**Pull request by jack-w-shaw**
It turns out we do not just disallow overlapping port ranges to be open across a singular unit, but across any units colocated on the same net node.

Fix UpdatePortRanges to do the stronger verification

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

Unit tests pass

### CODE REVIEW

File: domain/port/state/state_test.go [hunk 1]
```diff
@@ -498,6 +494,81 @@ func (s *stateSuite) TestUpdateUnitPortsClosePortRangeOverlapping(c *gc.C) {
...
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = st.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		return st.UpdateUnitPorts(ctx, s.unitUUID, network.GroupedPortRanges{"endpoint": {{Protocol: "tcp", FromPort: 2000, ToPort: 2250}}}, network.GroupedPortRanges{})
+	})
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+// NOTE (jack-w-shaw): This the behaviour this test is asserting is different to
```
Comment (SimonRichardson) [inline domain/port/state/state_test.go:531]: Re-write this, it doesn't make sense.

**Review by SimonRichardson — COMMENTED** [review]
I think this is OK, but I'd want @manadart to approve.

File: domain/port/state/state.go [hunk 2]
```diff
@@ -185,28 +185,33 @@ func (st *State) UpdateUnitPorts(
...
-		currentOpenedPorts, err := st.getUnitOpenedPorts(ctx, tx, unitUUID)
+		colocatedOpenedPorts, err := st.getColocatedOpenedPorts(ctx, tx, unitUUID)
 		if err != nil {
-			return errors.Annotatef(err, "getting opened ports for unit %q", unit)
+			return errors.Annotatef(err, "getting opened ports co-located with unit %q", unit)
 		}
 
 		allInputPortRanges := append(openPorts.UniquePortRanges(), closePorts.UniquePortRanges()...)
-		err = verifyNoPortRangeConflicts(currentOpenedPorts, allInputPortRanges)
+		err = verifyNoPortRangeConflicts(colocatedOpenedPorts, allInputPortRanges)
```
Comment (jack-w-shaw) [inline domain/port/state/state.go:194]: NOTE: In Mongo, we actually disallow equal port ranges being opened on co-located units, despite the fact we allow them on different endpoints of the same unit

So this behaviour here is slightly different. We will now allow equal opened port ranges on co-located units.

This is fine because downstream when we open port ranges we de-dup across endpoints. We will just do that against co-located units as well

**Review by manadart — COMMENTED** [review]
This looks correct, but I think you need to do a refactoring exercise.

Reading `UpdateUnitPorts` in the service layer, we have the reconciliation of open and close ranges, including wildcard checking, with `state.UpdateUnitPorts` called at the end.

The fact that this method uses `RunAtomic` leads the reader to believe that reading it endows them with the the knowledge of what business logic is considered in this method, But without reading `state.UpdateUnitPorts` this is false.

The service layer represents the state definition of `UpdateUnitPorts` and an _atomic_ operation, but this is not the case. We query all the ports for colocated units and verify that we have no conflicts, before ensuring that all the unit endpoints are present, getting the current open ports, and setting open/closed with those in consideration.

If the service definition of `UpdateUnitPorts` is not where all the business logic is applied, with the state methods being simple persistence parts, we've got the worst of it by using `RunAtomic`.

**Comment by jack-w-shaw** [comment]
> This looks correct, but I think you need to do a refactoring exercise.
> 
> Reading `UpdateUnitPorts` in the service layer, we have the reconciliation of open and close ranges, including wildcard checking, with `state.UpdateUnitPorts` called at the end.
> 
> The fact that this method uses `RunAtomic` leads the reader to believe that reading it endows them with the the knowledge of what business logic is considered in this method, But without reading `state.UpdateUnitPorts` this is false.
> 
> The service layer represents the state definition of `UpdateUnitPorts` and an _atomic_ operation, but this is not the case. We query all the ports for colocated units and verify that we have no conflicts, before ensuring that all the unit endpoints are present, getting the current open ports, and setting open/closed with those in consideration.
> 
> If the service definition of `UpdateUnitPorts` is not where all the business logic is applied, with the state methods being simple persistence parts, we've got the worst of it by using `RunAtomic`.

I guess it depends how you define 'business logic' & 'persistence logic' here. I initially designed my state + service layer to do just this, to keep business logic in the service layer, and then persistence logic in the state layer. Now that conflict-detecting has been changed slightly, I do agree a little business logic has found its way into the state layer.

i.e. the state layer now need to be aware of the specific fact that we disallow opened port ranges to conflict _on co-located units_. 

I agree this fact should be raised up into the service layer. Arguably, the fact we disallow overlapping port ranges at all is business logic. I do feel a good case can be made this is persistence logic, but it will probably have to come out anyway with this change.

I feel ensuring endpoints exist, and determining how to handle opened and closed ports (ignoring open ranges if they're already open, removing a port range if we're closing it, etc.) all counts as persistence logic, which is sensible to include within UpdateUnitPorts.

So, to be specific, the way I would resolve your concern would be to raise the calls to `getColocatedOpenedPorts` and `verifiyNoPortRangeConflicts` out of state + into the service.

@manadart what do you think?

File: domain/port/service/service.go [hunk 3]
```diff
@@ -232,5 +250,5 @@ func verifyNoPortRangeConflicts(openPorts, closePorts network.GroupedPortRanges)
 	if len(conflicts) == 0 {
 		return nil
 	}
-	return errors.NotValidf("conflicting port ranges: %s", conflicts)
+	return errors.Annotatef(port.ErrPortRangeConflict, "conflicting port ranges: %s", conflicts)
```
Comment (manadart) [inline domain/port/service/service.go:253]: Use `Errorf` as per Tom's email.

File: domain/port/state/state.go [hunk 4]
```diff
@@ -185,28 +228,22 @@ func (st *State) UpdateUnitPorts(
...
 
 	return domain.Run(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		currentOpenedPorts, err := st.getUnitOpenedPorts(ctx, tx, unitUUID)
-		if err != nil {
-			return errors.Annotatef(err, "getting opened ports for unit %q", unit)
-		}
-
-		allInputPortRanges := append(openPorts.UniquePortRanges(), closePorts.UniquePortRanges()...)
-		err = verifyNoPortRangeConflicts(currentOpenedPorts, allInputPortRanges)
+		endpoints, err := st.ensureEndpoints(ctx, tx, unitUUID, endpointsUnderAction)
```
Comment (manadart) [inline domain/port/state/state.go:231]: If you describe what this method does in plain English, it should be obvious what should be lifted into the service layer as orchestration, and what should remain in state as modules of retrieval/persistence.

**Review by manadart — APPROVED** [review]
This can land with a removal of the `Annotatef`. 

Reworking for correct separation can come later, as this patch puts its change in the correct place.

File: domain/port/service/service.go (hunk 3 above)
Comment (jack-w-shaw) [reply domain/port/service/service.go]: Done
